drivers: led: Add lp5569 led controller driver

jonas-rem opened the pull request:
Add a minimal driver and sample for the ti lp5569 led controller. The driver supports multiple instances. Commands on|off|set_brightness are supported.

In addition this PR adds the possibility for the driver to use device power management with or without e.g. using a power domain for power gating the led controller device.

Commit:
samples: drivers: led: add a sample for ti lp5569

Add a sample for the ti lp5569 led controller driver.

Signed-off-by: Jonas Remmert <j.remmert@phytec.de>

Commit:
tests: driver: build_all: add lp5569 to tests

Add LED controller driver to list of tests.

Signed-off-by: Jonas Remmert <j.remmert@phytec.de>

simonguinot commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,169 @@
+static inline int lp5569_led_on(const struct device *dev, uint32_t led)
+{
+	const struct lp5569_config *config = dev->config;
+
```
Eventually you should make it a call to `lp5569_led_set_brightness`. This would avoid some code duplication.

simonguinot commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,169 @@
+static inline int lp5569_led_off(const struct device *dev, uint32_t led)
+{
+	const struct lp5569_config *config = dev->config;
+
```
Same here.

simonguinot commented on samples/drivers/led_lp5569/src/main.c:
```diff
@@ -0,0 +1,62 @@
+			}
+
+			k_sleep(DELAY_TIME);
+		}
```
Please can you add a breathing test as well ? To demonstrate the `set_brightness` API function.

jremmert-phytec-iot replied:
Good point! I changed the sample so that it enables all leds step-by-step and then slowly dims them simultaneously until off.

simonguinot reviewed (changes requested):
Hi @jremmert-phytec-iot,

Thanks again for this driver. It is looking good overall.

Please find my requests, comments and questions below.

jremmert-phytec-iot commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,169 @@
+	};								\
+	static struct lp5569_data lp5569_data_##id;			\
+									\
+	DEVICE_DT_INST_DEFINE(id, &lp5569_init_dev,			\
```
fixed, first commit now builds without issues.

jremmert-phytec-iot commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,169 @@
+
+struct lp5569_data {
+	struct led_data dev_data;
+};
```
Removed and replaced with `NULL` in `DEVICE_DT_INST_DEFINE`

jremmert-phytec-iot commented on drivers/led/lp5569.c:
```diff
@@ -146,6 +147,46 @@ static int lp5569_init(const struct device *dev)
+	switch (action) {
+	case PM_DEVICE_ACTION_TURN_ON:
+	case PM_DEVICE_ACTION_RESUME:
+		lp5569_enable(dev);
```
fixed

jremmert-phytec-iot commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,169 @@
+	}
+
+	/* Map 0-100 % to 0-255 pwm register value */
+	val = brightness * 2.55;
```
changed it to `255 / 100`

jremmert-phytec-iot commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,169 @@
+#define LP5569_POWERSAVE_EN		(1 << 5)
+
+/* PWM Registers for controlling the duty-cycle */
+#define LP5569_LED0_PWM			0x16
```
Changed it. I added them initially to give an overview over important driver registers, but I guess less code for the same function has more advantages.

jremmert-phytec-iot commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,169 @@
+#define LP5569_LED5_CURRENT		0x27
+#define LP5569_LED6_CURRENT		0x28
+#define LP5569_LED7_CURRENT		0x29
+#define LP5569_LED8_CURRENT		0x2A
```
Nothing to configure, this drivers sticks with the reset values of the lp5569 device. I dropped them.

jremmert-phytec-iot commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,169 @@
+
+struct lp5569_config {
+	struct i2c_dt_spec bus;
+	const struct device *vin_supply;
```
fixed

Commit:
drivers: led: Add lp5569 led controller driver

Add a minimal driver for the ti lp5569 led controller. The driver supports
multiple instances. Commands on|off|set_brightness are supported.

Signed-off-by: Jonas Remmert <j.remmert@phytec.de>

Commit:
drivers: led: add power management for ti lp5569

Adding device power management for the ti lp5569 led controller driver.
Allow the option to either use device power managemnt while keeping the
voltage on or to use device power management including using e.g. a voltage
domain that controls the led controller voltage on-demand.

Signed-off-by: Jonas Remmert <j.remmert@phytec.de>

jremmert-phytec-iot commented:
@simonguinot thanks for your review! I added all your suggestions.

simonguinot reviewed (dismissed):
It looks good to me @jremmert-phytec-iot ! Thanks.

simonguinot commented:
Hi @Mani-Sadhasivam. It is good for me. Please have a look.

simonguinot commented:
And let's add more eyes in order to get one more review.

bbilas commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,165 @@
+{
+	const struct lp5569_config *config = dev->config;
+
+	if (!device_is_ready(config->bus.bus)) {
```
```suggestion
	if (!i2c_is_ready_dt(&config->bus)) {
```

jremmert-phytec-iot replied:
fixed it

bbilas commented on drivers/led/Kconfig.lp5569:
```diff
@@ -0,0 +1,13 @@
+	select I2C
+	help
+	  Enable LED driver for LP5569.
+
```
Extra nl.

bbilas commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,172 @@
+	/* Map 0-100 % to 0-255 pwm register value */
+	val = brightness * 255 / 100;
+
+
```
Extra nl.

bbilas commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,172 @@
+
+
+	ret = i2c_reg_write_byte_dt(&config->bus, LP5569_LED0_PWM + led, val);
+	if (ret) {
```
Could you please change it to `if (ret < 0)` everywhere?

jremmert-phytec-iot replied:
`i2c_reg_write_byte_dt` returns `i2c_reg_write_byte` which specifies `0` or `-EIO` as return values. Doesn't `ret` cover more cases than `ret < 0`?

jremmert-phytec-iot replied:
I did some reading and found that your suggestion seems to be the better approach. When considering safety and extensibility it's better to be explicit and specific. I will change it later today.

bbilas replied:
Thanks, let me know when it will be done.

jremmert-phytec-iot replied:
@bbilas just fixed all points.

bbilas commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,171 @@
+#include <zephyr/pm/device.h>
+#include <zephyr/kernel.h>
+
+#define LOG_LEVEL CONFIG_LED_LOG_LEVEL
```
You can remove that and use `LOG_MODULE_REGISTER(lp5569, CONFIG_LED_LOG_LEVEL);`

bbilas commented on drivers/led/lp5569.c:
```diff
@@ -0,0 +1,171 @@
+#define LP5569_CHIP_EN			(1 << 6)
+
+#define LP5569_MISC			0x2F
+#define LP5569_POWERSAVE_EN		(1 << 5)
```
Why not `BIT` macro?

jremmert-phytec-iot replied:
fixed

bbilas commented on samples/drivers/led_lp5569/src/main.c:
```diff
@@ -0,0 +1,65 @@
+#include <zephyr/drivers/led.h>
+#include <zephyr/kernel.h>
+
+#define LOG_LEVEL CONFIG_LOG_DEFAULT_LEVEL
```
ditto

jremmert-phytec-iot replied:
fixed

bbilas reviewed (commented):
3 minor comments, otherwise LGTM!
